libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+ * cuda::memcpy_async dispatch helper functions
+ *
+ * - __get_size_align struct to determine the alignment from a size type.
+ ***********************************************************************/
 
-    NV_DISPATCH_TARGET(
-        NV_PROVIDES_SM_80,
-
-        if (__isShared(__destination) && __isGlobal(__source)) {
-            if (_Native_alignment < 4) {
-                auto __source_address = reinterpret_cast<_CUDA_VSTD::uintptr_t>(__source);
-                auto __destination_address = reinterpret_cast<_CUDA_VSTD::uintptr_t>(__destination);
-
-                // Lowest bit set will tell us what the common alignment of the three values is.
-                auto _Alignment = __ffs(__source_address | __destination_address | __size);
-
-                switch (_Alignment) {
-                    default:
-                        __memcpy_async_sm_dispatch<_Native_alignment>(__group, __destination, __source, __size, __sync, __is_async);
-                    case 4: __is_async = __memcpy_async_impl<8>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 3: __is_async = __memcpy_async_impl<4>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 2: // fallthrough
-                    case 1: __is_async = __memcpy_async_impl<1>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                }
-            }
-            else {
-                __is_async = __memcpy_async_impl<_Native_alignment>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size());
+// The __get_size_align struct provides a way to interrogate the guaranteed
+// "alignment" of a provided size. In this case, an n-byte aligned size means
+// that the size is a multiple of n.
+//
+// Use as follows:
+// static_assert(__get_size_align<size_t>::align == 1)
+// static_assert(__get_size_align<aligned_size_t<n>>::align == n)
+template< class... >
+using __custom_void_t = void;
+
+// Default impl: always returns 1.
+template <typename, typename = void>
+struct __get_size_align {
+    static constexpr int align = 1;
+};
+
+// aligned_size_t<n> overload: return n.
+template <typename T>
+struct __get_size_align<T, __custom_void_t<decltype(T::align)>> {
+    static constexpr int align = T::align;
+};
+
+/***********************************************************************
+ * cuda::memcpy_async dispatch
+ *
+ * The dispatch mechanism takes all the arguments and dispatches to the
+ * fastest asynchronous copy mechanism available.
+ *
+ * It returns a __completion_mechanism that indicates which completion mechanism
+ * was used by the copy mechanism. This value can be used by the sync object to
+ * further synchronize if necessary.
+ *
+ ***********************************************************************/
+
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
```

Comment:
ditto nodiscard
```suggestion
_LIBCUDACXX_NODISCARD_ATTRIBUTE inline __device__
```

Reply:
Applied nodiscard attribute in offline commit.